deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,363 @@
+        time.sleep(cfg.timeout1)
+
+    print(f"bpfiles = {bpfiles}")
+
+    # Wait for enough time steps in each bp file
+    while(True):
+        enough = True
+        for bp in bpfiles:
+            com = f"bpls {bp}"
+            a = subprocess.getstatusoutput(com)
+            if(a[0] != 0):
+                enough = False
+                print(f"Waiting, a = {a}, {bp}")
+                break
+            try:
+                steps = int(a[1].split("\n")[0].split("*")[0].split(" ")[-1])
+            except:
+                steps = 0
+                enough = False
+            if(steps < cfg.min_step_increment):
+                enough = False
+                print(f"Waiting, steps = {steps}, {bp}")
+                break
+        if(enough):
+            break
+        else:
+            time.sleep(cfg.timeout2)
+
+    return bpfiles
+
+
+def dirs(cfg):
+    top_dir = cfg.output_path
+    tmp_dir = f"{top_dir}/tmp"
+    published_dir = f"{top_dir}/published_outliers"
+
+    if(not os.path.exists(tmp_dir)):
+         os.mkdir(tmp_dir)
+    if(not os.path.exists(published_dir)):
+         os.mkdir(published_dir)
+    return top_dir, tmp_dir, published_dir
+
+def predict(cfg, model_path, cvae_input, batch_size=32):
+    cvae = build_model(cfg, model_path)
+    input = np.expand_dims(cvae_input[0], axis = -1)
+
+    cm_predict = cvae.return_embeddings(input, batch_size)
+    del cvae 
+    K.clear_session()
+    return cm_predict
+
+def outliers_from_latent(cm_predict, eps=0.35, min_samples=10):
+    cm_predict = cp.asarray(cm_predict)
+    db = DBSCAN(eps=eps, min_samples=min_samples, max_mbytes_per_batch=100).fit(cm_predict)
+    db_label = db.labels_.to_array()
+    print("unique labels = ", np.unique(db_label))
+    outlier_list = np.where(db_label == -1)
+    K.clear_session()
+    return outlier_list
+
+def cluster(cfg, cm_predict, outlier_list, eps, min_samples):
+    outlier_count = cfg.outlier_count
+    while outlier_count > 0:
+        n_outlier = 0
+        try:
+            outliers = np.squeeze(outliers_from_latent(cm_predict, eps=eps, min_samples=min_samples)) 
+            n_outlier = len(outliers)
+        except Exception as e:
+            print(e)
+            print("No outliers found")
+
+        print(f'eps = {eps}, min_samples = {min_samples}, number of outlier found: {n_outlier}')
+
+        if n_outlier > cfg.outlier_max: 
+            eps = eps + 0.09*random.random()
+            min_samples -= int(random.random() < 0.5)
+            min_samples = max(5, min_samples)
+        elif n_outlier < cfg.outlier_min:
+            eps = max(0.01, eps - 0.09*random.random())
+            min_samples += int(random.random() < 0.5)
+        else: 
+            outlier_list.append(outliers) 
+            break 
+        outlier_count -= 1
+    return eps, min_samples
+
+
+def write_pdb_frame(frame, original_pdb, output_pdb_fn):
+    pdb = PDBFile(str(original_pdb))
+    f = open(str(output_pdb_fn), 'w')
```

Comment:
Can use a context manager here

Reply:
Not sure I understand

Reply:
```
with open( str(output_pdb_fn), 'w'):
   ...
```
Just a small change.

Reply:
Done. 
I did not know this construction is called "context manager".